File: java/interpreter.java
import java.util.Scanner;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Arrays;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.time.LocalDate;
import java.time.LocalTime;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import javax.script.ScriptEngineManager;
import javax.script.ScriptEngine;
import java.lang.Math;
import javax.vecmath.Vector3d;
public class interpreter{
  //static ArrayList<String> custom = ArrayList<String>();
  static int line;
  static ArrayList<goldObject> vars = ArrayList<goldObject>();
  static errorHandler Ehandler = errorHandler();
  static String inloc;
  static String ret;
  static String[] prevars;
  static Scanner inputHandler = new Scanner(System.in);
  static ScriptEngineManager mgr = new ScriptEngineManager();
  static ScriptEngine js = mgr.getEngineByName("JavaScript");
  public static String parseList(int[] a){
    int b = 1;
    int c = 1;
    String d = String.valueOf(int[0]);
    while(b==1){
      try{
        d=concate(d,concate("#",String.valueOf(int[c])));
        c++;
      } catch(Exception e){
        b=0;
      }
    }
  }
  public static String parsetime(String so){
    String s1;
    if(so.equals("%x")){
      s1 = "MM-dd-yyyy";
    } else if(so.equals("%d")){
      s1 = "dd";
    } else if(so.equals("%H")){
      s1 = "HH";
    } else if(so.equals("%M")){
      s1 = "mm";
    } else if(so.equals("%m")){
      s1 = "MM";
    } else if(so.equals("%f")){
      s1 = "ns";
    } else if(so.equals("%S")){
      s1 = "ss";
    } else if(so.equals(%Y")){
      s1 = "yyyy";
    } else if(so.equals("%y")){
      s1 = "yy";
    } else if(so.equals("%X")){
      s1 = "HH:mm:ss";
    } else{
      this.Ehandler.adderror(new goldException("Gold_Interpreter:Error[1]:Failed to parse time formatting","System_error"));
    }
    return s1;
  }
  public static String concate(Object o0,Object o1){
    return o0+""+o1;
  }
  public static interpreter(String[] args){
    this.inloc = args[0];
  }
  public static void gocmd(String args){
    try{
      this.line = Integer.parseInt(args[0]);
    } catch(Exception e){
      this.Ehandler.adderror(new goldException(concate(this.inloc,":Error[1]:Failed to parse command <go>"),"Command_error"));
    }
  }
  public static void goifcmd(String[] args){
    try{
      int a = Integer.parseInt(args[0]);
      String b = args[1];
      Float c = this.vars.get(Integer.parseInt(args[2])).getValue();
      Float d = this.vars.get(Integer.parseInt(args[3])).getValue();
      if (b.equals("==")){
        if(c==d){
          this.line = a;
        }
      } else if(b.equals(">")){
        if(c>d){
          this.line = a;
        }
      } else if(b.equals("<")){
        if(c<d){
          this.line = a;
        }
      } else if(b.equals(">=")){
        if(c>=d){
          this.line = a;
        }
      } else if(b.equals("<=")){
        if(c<=d){
          this.line = a;
        }
      } else if(b.equals("!")){
        if(c!=d){
          this.line = a;
        }
      } else{
        this.Ehandler.adderror(new goldException(concate(this.inloc,":Error[2]:Failed to parse command <goif>"),"Command_error"));
      }
    } catch(Exception e){
      this.Ehandler.adderror(new goldException(concate(this.inloc,":Error[2]:Failed to parse command <goif>"),"Command_error"));
    }
  }
  public static void throwcmd(String[] args){
    String a;
    String b;
    try{
      a = args[0];
      try{
        b = args[1];
        if(b.equals("Command_error")||b.equals("System_error")){
          b="Thrown_error";
        }
      } catch(Exception e){
        b="Thrown_error";
      }
    } catch(Exception e){
      a="Error[?]:undefined error";
      b="Thrown_error";
    }
    this.Ehandler.adderror(new goldException(concate(this.inloc,concate(":",a)),b));
  }
  public static void importcmd(String[] args){
    try{
      Interpreter tempinter = new Interpreter({args[2]});
      vars.set(Integer.parseInt(args[1]),tempinter.run(args[0],this.vars.get(Integer.parseInt(args[3])).getValue().split("#")));
    } catch(Exception e){
      this.Ehandler.adderror(new goldException(concate(this.inloc,":Error[3]:Failed to parse command <import>"),"Command_error"));
    }
  }
  public static void returncmd(String[] args){
    try{
      String a = args[0];
      if(a.equals("String")){
        this.ret=args[1];
      } else if(a.equals("Variable"){
        this.ret=this.vars.get(Integer.parseInt(args[1])).asString();
      }
    } catch(Exception e){
      this.Ehandler.adderror(new goldException(concate(this.inloc,":Error[4]:Failed to parse command <return>"),"Command_error"));
    }
  }
  public static void assigncmd(String[] args){
    try{
      String a = args[0];
      if(a.equals("String")){
        this.vars.set(Integer.parseInt(args[1]),new goldString(args[2]));
      } else if(a.equals("Prevar")){
        a=args[1]
        if(a.equals("String")){
          this.vars.set(Integer.parseInt(args[2]),new goldString(this.prevars[Integer.parseInt(args[3])]));
        } else if(a.equals("Number")){
          this.vars.set(Integer.parseInt(args[2]),new goldNum(Float.valueFloat(this.prevars[Integer.parseInt(args[3])])));
        }
      } else if(a.equals("Number")){
        this.vars.set(Integer.parseInt(args[1]),new goldNum(Float.valueFloat(args[3])));
      } else if(a.equals("Lorem_Ipsum")){
        this.vars.set(Integer.parseInt(args[1]),new goldString("We Three Kings of Orient are,\nBearing gifts we traverse afar,\nField and fountain,\nMoor and mountain,\nFollowing yonder Star."));
      } else{
        this.Ehandler.adderror(new goldException(concate(this.inloc,":Error[5]:Failed to parse command <assign>"),"Command_error"));
      }
    } catch(Exception e){
      this.Ehandler.adderror(new goldException(concate(this.inloc,":Error[5]:Failed to parse command <assign>"),"Command_error"));
    }
  }
  public static void timecmd(String[] args){
    try{
      this.vars.set(Integer.parseInt(args[0]),new goldString(LocalDateTime.now(DateTimeFormatter.ofPattern(parsetime(args[1])))));
    } catch(Exception e){
      this.Ehandler.adderror(new goldException(concate(this.inloc,":Error[6]:Failed to parse command <time>"),"Command_error"));
    }
  }
  public static void outcmd(String[] args){
    try{
      a = args[0];
      if(a.equals("String")){
        System.out.print(args[1]);
      } else if(a.equals("Variable")){
        System.out.print(this.vars.get(Integer.parseInt(args[1])).getValue());
      } else if(a.equals("New")){
        System.out.println();
      } else if(a.equals("Return")){
        System.out.print("\r");
      } else{
        this.Ehandler.adderror(new goldException(concate(this.inloc,":Error[7]:Failed to parse command <out>"),"Command_error"));
      }
    } catch(Exception e){
      this.Ehandler.adderror(new goldException(concate(this.inloc,":Error[7]:Failed to parse command <out>"),"Command_error"));
    }
  }
  public static void implementcmd(String[] args){
    try{
      this.vars.set(Integer.parseInt(args[0]),new goldString("0"));
    } catch(Exception e){
      this.Ehandler.adderror(new goldException(concate(this.inloc,":Error[8]:Failed to parse command <implement>"),"Command_error"));
    }
  }
  public static void inputcmd(String[] args){
    try{
      this.vars.set(Integer.parseInt(args[0]),new goldString(this.inputHandler.nextLine()));
    } catch(Exception e){
      this.Ehandler.adderror(new goldException(concate(this.inloc,":Error[9]:Failed to parse command <input>"),"Command_error"));
    }
  }
  public static void operationcmd(String[] args){
    try{
      a = args[0];
      if(a.equals("+")){
        this.vars.set(Integer.parseInt(args[1]),new goldNum(this.vars.get(Integer.parseInt(args[2])).getValue()+this.vars.get(Integer.parseInt(args[3])).getValue()));
      } else if(a.equals("-")){
        this.vars.set(Integer.parseInt(args[1]),new goldNum(this.vars.get(Integer.parseInt(args[2])).getValue()-this.vars.get(Integer.parseInt(args[3])).getValue()));
      } else if(a.equals("*")){
        this.vars.set(Integer.parseInt(args[1]),new goldNum(this.vars.get(Integer.parseInt(args[2])).getValue()*this.vars.get(Integer.parseInt(args[3])).getValue()));
      } else if(a.equals(":")){
        this.vars.set(Integer.parseInt(args[1]),new goldNum(this.vars.get(Integer.parseInt(args[2])).getValue()/this.vars.get(Integer.parseInt(args[3])).getValue()));
      } else if(a.equals("**")){
        this.vars.set(Integer.parseInt(args[1]),new goldNum(Math.pow(this.vars.get(Integer.parseInt(args[2])).getValue(),this.vars.get(Integer.parseInt(args[3])).getValue())));
      } else if(a.equals("::")){
        this.vars.set(Integer.parseInt(args[1]),new goldNum(Math.pow(this.vars.get(Integer.parseInt(args[2])).getValue(),(1f/this.vars.get(Integer.parseInt(args[3])).getValue()))));
      } else if(a.equals("%")){
        this.vars.set(Integer.parseInt(args[1]),new goldNum(this.vars.get(Integer.parseInt(args[2])).getValue()%this.vars.get(Integer.parseInt(args[3])).getValue()));
      } else if(a.equals("$")){
        this.vars.set(Integer.parseInt(args[1]),new goldNum(((this.vars.get(Integer.parseInt(args[2])).getValue())%2f==0f)?0f:1f));
      } else if(a.equals("&&")){
        this.vars.set(Integer.parseInt(args[1]),new goldString(concate(this.vars.get(Integer.parseInt(args[2])).getValue(),this.vars.get(Integer.parseInt(args[3])).getValue())));
      } else if(a.equals("&")){
        this.vars.set(Integer.parseInt(args[1]),new goldNum(this.vars.get(Integer.parseInt(args[2])).getValue()&this.vars.get(Integer.parseInt(args[3])).getValue()));
      } else if(a.equals("|")){
        this.vars.set(Integer.parseInt(args[1]),new goldNum(this.vars.get(Integer.parseInt(args[2])).getValue()|this.vars.get(Integer.parseInt(args[3])).getValue()));
      } else if(a.equals("^")){
        this.vars.set(Integer.parseInt(args[1]),new goldNum(this.vars.get(Integer.parseInt(args[2])).getValue()^this.vars.get(Integer.parseInt(args[3])).getValue()));
      } else if(a.equals("~")){
        this.vars.set(Integer.parseInt(args[1]),new goldNum(~this.vars.get(Integer.parseInt(args[2])).getValue()));
      } else if(a.equals(">>")){
        this.vars.set(Integer.parseInt(args[1]),new goldNum(this.vars.get(Integer.parseInt(args[2])).getValue()>>));
      } else if(a.equals(">>>")){
        this.vars.set(Integer.parseInt(args[1]),new goldNum(this.vars.get(Integer.parseInt(args[2])).getValue()>>>));
      } else if(a.equals("<<")){
        this.vars.set(Integer.parseInt(args[1]),new goldNum(this.vars.get(Integer.parseInt(args[2])).getValue()<<));
      } else if(a.equals("E")){
        this.vars.set(Integer.parseInt(args[1]),new goldString(this.js.eval(this.vars.get(Integer.parseInt(args[2])).getValue())));
      } else if(a.equals("N")){
        this.vars.set(Integer.parseInt(args[1]),new goldNum(Float.valueFloat(this.vars.get(Integer.parseInt(args[2])))));
      } else if(a.equals("X")){
        this.vars.set(Integer.parseInt(args[1]),new goldString(parseList((new d3vector()).cross((new d3vector(this.vars.get(Integer.parseInt(args[2])).getValue().split("#")).to_list()),(new d3vector(this.vars.get(Integer.parseInt(args[3])).getValue().split("#")).to_list())).to_list())));
      } else if(a.equals("D")){
        this.vars.set(Integer.parseInt(args[1]),new goldString(parseList((new d3vector()).
      } else{
        this.Ehandler.adderror(new goldException(concate(this.inloc,":Error[10]:Failed to parse command <operation>"),"Command_error"));
      }
    } catch(Exception e){
      this.Ehandler.adderror(new goldException(concate(this.inloc,":Error[10]:Failed to parse command <operation>"),"Command_error"));
    }
  }
  public static Object run(String filename,String[] prevars){
    try{
      file_reader fileHandler = new file_reader(concate(filename,".gold"));
      this.prevars = prevars;
      ArrayList<String> contents = new ArrayList<String>();
      contents.addAll(fileHandler.read());
      int a = 1;
      int b = 0;
      String tempstring;
      ArrayList<String> command = new ArrayList<String>();
      String[] templist;
      iny c = 0;
      while(c<250){
        this.vars.add(new goldObject);
        c+=1
      }
      while(a==1){
        try{
          command.addAll(Arrays.asList(content.get(b).split(" ")));
          tempstring = command.get(0);
          command.remove(0);
          if(tempstring.equals("jgo")){
            this.gocmd(command.toArray());
          } else if(tempstring.equals("goif")){
            this.goifcmd(command.toArray());
          } else if(tempstring.equals("throw")){
            this.throwcmd(command.toArray());
          } else if(tempstring.equals("import")){
            this.importcmd(command.toArray());
          } else if(tempstring.equals("return")){
            this.returncmd(command.toArray());
          } else if(tempstring.equals("assign")){
            this.returncmd(command.toArray());
          } else if(tempstring.equals("time")){
            this.returncmd(command.toArray());
          } else if(tempstring.equals("out")){
            this.outcmd(command.toArray());
          } else if(tempstring.equals("implement")){
            this.implementcmd(command.toArray());
          } else if(tempstring.equals("operation")){
            this.operationcmd(command.toArray());
          } else if(tempstring.equals("comment")){
            System.out.print();
          } else{
            this.Ehandler.adderror(new goldException(concate(this.inloc,":Error[0]:Unknown command"),"Command_error"));
          }
          command.clear();
          b += 1;
        } catch{
          a = 0;
        }
      }
    } catch(Exception e){
      this.Ehandler.adderror(new goldException("Gold_Interpreter:Error[0]:Critical failure","System_error"));
    }
    this.Ehandler.rethrow();
    return this.ret;
  }
}

File: interpreter.py
import datetime
import random
import js2py
class goldException():
    def __init__(self,error,error_type):
        self.message="\033[31m"+error+"\033[0m"
        self.error_type=error_type
    def geterror(self):
        return self.message
    def gettype(self):
        return self.error_type
class errorHandler():
    def __init__(self):
        self.errors=[]
        self.nexterror=0
    def adderror(self,error):
        self.errors+=[error]
        self.nexterror+=1
    def rethrow(self):
        a=1
        b=""
        c=0
        while a==1:
            try:
                d=self.errors[c]
                b=d.getmessage()
                b+=","
                b+=d.gettype()
                print(b)
                c+=1
            except:
                a=0
class d3_vector():
    def __init__(self,a=0,b=0,c=0):
        self.a=a
        self.b=b
        self.c=c
    def from_list(self,a):
        self.a=int(a[0])
        self.a=int(a[1])
        self.a=int(a[2])
    def cross(self,a,b):
        self.a=a[1]*b[2]-a[2]*b[1]
        self.b=a[2]*b[0]-a[0]*b[2]
        self.c=a[0]*b[1]-a[1]*b[0]
    def to_list(self):
        a=[self.a,self.b,self.c]
        return a
def concate(s0,s1):
  return s0+""+s1
def parse_list(a):
    d=1
    b=1
    c=str(a[0])
    while b==1:
        try:
            c=concate(c,concate("#",a[d]))
            d+=1
        except:
            b=0
    return c
class interpreter():
    def __init__(self,args):
        self.line=0
        self.inloc=args[0]
        self.vars=[]
        self.ret=""
        self.prevars=[]
        self.vars=[]
        self.Ehandler=errorHandler()
    def gocmd(self,args):
        try:
            self.line=int(args[0])
        except:
            self.Ehandler.adderror(goldException(concate(self.inloc,":Error[1]:Failed to parse command <go>"),"Command_error"))
    def goifcmd(self,args):
        try:
            if args[1]=="==":
                if self.vars[int(args[2])]==self.vars[int(args[3])]:
                    self.line=int(args[0])
            elif args[1]==">":
                if self.vars[int(args[2])]>self.vars[int(args[3])]:
                    self.line=int(args[0])
            elif args[1]=="<":
                if self.vars[int(args[2])]<self.vars[int(args[3])]:
                    self.line=int(args[0])
            elif args[1]==">=":
                if self.vars[int(args[2])]>=self.vars[int(args[3])]:
                    self.line=int(args[0])
            elif args[1]=="<=":
                if self.vars[int(args[2])]<=self.vars[int(args[3])]:
                    self.line=int(args[0])
            elif args[1]=="!=":
                if self.vars[int(args[2])]!=self.vars[int(args[3])]:
                    self.line=int(args[0])
            else:
                self.Ehandler.adderror(goldException(concate(self.inloc,":Error[2]:Failed to parse command <goif>"),"Command_error"))
        except:
            self.Ehandler.adderror(goldException(concate(self.inloc,":Error[2]:Failed to parse command <goif>"),"Command_error"))
    def throwcmd(self,args):
        try:
            self.Ehandler.adderror(goldException(concate(self.inloc,concate(":",args[0])),(args[1] if (args[1]=="Command_error" or args[1]=="System_error") else "Thrown_error")))
        except:
            self.Ehandler.adderror(goldException(concate(self.inloc,":Error[?]:Undefined"),"Thrown_error"))
    def importcmd(self,args):
        try:
            self.vars[int(args[1])]=(Interpreter(args[2].split("#"))).run(args[0],(self.vars[int(args[3])]).split("#"))
        except:
            self.Ehandler.adderror(goldException(concate(self.inloc,":Error[3]:Failed to parse command <import>"),"Command_error"))
    def returncmd(self,args):
        try:
            if args[0]=="String":
                self.ret=args[1]
            elif args[0]=="Variable":
                self.ret=str(self.vars[int(args[1])])
            else:
                self.Ehandler.adderror(goldException(concate(self.inloc,":Error[4]:Failed to parse command <return>"),"Command_error"))
        except:
            self.Ehandler.adderror(goldException(concate(self.inloc,":Error[4]:Failed to parse command <return>"),"Command_error"))
    def assigncmd(self,args):
        try:
            if args[0]=="String":
                self.vars[int(args[1])]=args[2]
            elif args[0]=="Prevar":
                if args[1]=="String":
                    self.vars[int(args[2])]=self.prevars[int(args[3])]
                elif args[1]=="Number":
                    self.vars[int(args[2])]=float(self.prevars[int(args[3])])
            elif args[0]=="Number":
                self.vars[int(args[1])]=float(args[2])
            else:
                self.Ehandler.adderror(goldException(concate(self.inloc,":Error[5]:Failed to parse command <assign>"),"Command_error"))
        except:
            self.Ehandler.adderror(goldException(concate(self.inloc,":Error[5]:Failed to parse command <assign>"),"Command_error"))
    def timecmd(self,args):
        try:
            self.vars[int(args[0])]=(datetime.datetime.now()).strftime(args[1])
        except:
            self.Ehandler.adderror(goldException(concate(self.inloc,":Error[6]:Failed to parse command <time>"),"Command_error"))
    def outcmd(self,args):
        try:
            if args[0]=="String":
                print(args[1],end="")
            elif args[0]=="Variable":
                print(self.vars[int(args[1])],end="")
            elif args[0]=="New":
                print()
            elif args[0]=="Return":
                print("\r",end="")
            else:
                self.Ehandler.adderror(goldException(concate(self.inloc,":Error[7]:Failed to parse command <out>"),"Command_error"))
        except:
            self.Ehandler.adderror(goldException(concate(self.inloc,":Error[7]:Failed to parse command <out>"),"Command_error"))
    def implementcmd(self,args):
        try:
            self.vars[int(args[0])]="1"
        except:
            self.Ehandler.adderror(goldException(concate(self.inloc,":Error[8]:Failed to parse command <implement>"),"Command_error"))
    def inputcmd(self,args):
        try:
            self.vars[int(args[0])]=input()
        except:
            self.Ehandler.adderror(goldException(concate(self.inloc,":Error[9]:Failed to parse command <input>"),"Command_error"))
    def operationcmd(self,args):
        try:
            if args[0]=="+":
                self.vars[int(args[1])]=self.vars[int(args[2])]+self.vars[int(args[3])]
            elif args[0]=="-":
                self.vars[int(args[1])]=self.vars[int(args[2])]-self.vars[int(args[3])]
            elif args[0]=="*":
                self.vars[int(args[1])]=self.vars[int(args[2])]*self.vars[int(args[3])]
            elif args[0]==":":
                self.vars[int(args[1])]=self.vars[int(args[2])]/self.vars[int(args[3])]
            elif args[0]=="**":
                self.vars[int(args[1])]=self.vars[int(args[2])]**self.vars[int(args[3])]
            elif args[0]=="::":
                self.vars[int(args[1])]=self.vars[int(args[2])]**(1/self.vars[int(args[3])])
            elif args[0]=="%":
                self.vars[int(args[1])]=self.vars[int(args[2])]%self.vars[int(args[3])]
            elif args[0]=="$":
                self.vars[int(args[1])]=(1 if (self.vars[int(args[2])]%2) else 0)
            elif args[0]=="&&":
                self.vars[int(args[1])]=concate(self.vars[int(args[2])],self.vars[int(args[3])])
            elif args[0]=="&":
                self.vars[int(args[1])]=self.vars[int(args[2])]&self.vars[int(args[3])]
            elif args[0]=="|":
                self.vars[int(args[1])]=self.vars[int(args[2])]|self.vars[int(args[3])]
            elif args[0]=="^":
                self.vars[int(args[1])]=self.vars[int(args[2])]^self.vars[int(args[3])]
            elif args[0]==">>":
                self.vars[int(args[1])]=self.vars[int(args[2])]>>self.vars[int(args[3])]
            elif args[0]==">>>":
                self.vars[int(args[1])]=self.vars[int(args[2])]>>self.vars[int(args[3])]
            elif args[0]=="<<":
                self.vars[int(args[1])]=self.vars[int(args[2])]<<self.vars[int(args[3])]
            elif args[0]=="~":
                self.vars[int(args[1])]=~(self.vars[int(args[2])])
            elif args[0]=="X":
                self.vars[int(args[1])]=d3vector().cross((d3vector().from_list((self.vars[int(args[2])]).split("#"))),(d3vector().from_list((self.vars[int(args[3])]).split("#"))))
            elif args[0]=="N":
                self.vars[int(args[1])]=float(self.vars[int(args[2])])
            elif args[0]=="E":
                funct=j2py.eval_js(self.vars[int(args[2])])
                self.vars[int(args[1])]=funct()
            else:
                self.Ehandler.adderror(goldException(concate(self.inloc,":Error[10]:Failed to parse command <operation>"),"Command_error"))
        except:
            self.Ehandler.adderror(goldException(concate(self.inloc,":Error[10]:Failed to parse command <operation>"),"Command_error"))
    def run(self,filename,prevars):
        try:
            a=1
            self.prevars=prevars
            fileHandler=open(concate(file_name,".gold"))
            contents=(fileHandler.read()).split("\n")
            b=0
            c=0
            while c<250:
                self.vars+=[""]
                c+=1
            while a==1:
                try:
                    command=contents[b]
                    d=command[0]
                    del command[0]
                    if d=="go":
                        self.gocmd(command)
                    elif d=="goif":
                        self.goifcmd(command)
                    elif d=="throw":
                        self.throwcmd(command)
                    elif d=="import":
                        self.importcmd(command)
                    elif d=="return":
                        self.returncmd(command)
                    elif d=="assign":
                        self.assigncmd(command)
                    elif d=="time":
                        self.timecmd(command)
                    elif d=="out":
                        self.outcmd(command)
                    elif d=="implement":
                        self.implementcmd(command)
                    elif d=="input":
                        self.inputcmd(command)
                    elif d=="operation":
                        self.operationcmd(command)
                    elif d=="comment":
                        print("",end="")
                    else:
                        self.Ehandler.adderror(goldException(concate(self.inloc,":Error[0]:Unknown command"),"Command_error"))
                    b+=1
                except:
                    a=0
        except:
            self.Ehandler.adderror(goldException("Gold_Interpreter:Error[0]:Critical failure","System_error"))
